DIPs/DIP1004.md

```diff
@@ -0,0 +1,272 @@
+{
+    int sum;
+
+    this ( int x, int y )
+    {
+        this.sum = x + y;
+    }
+}
+
+class Derived : Base
+{
+    this ( )
+    {
+        super(5, 5);
+    }
+}
+
+void main ( )
+{
+    auto d = new Derived(10, 10);  // error
+}
+```
+
+The rationale for this current limitation: If the `Derived` object was constructible
+by calling only the `Base` class constructor then the `Derived` object itself
+could be left in an *unexpected* state. This is because the `Derived` constructors could
+be skipped and never called and potentially leaving `Derived`'s own internal fields
+in an unexpected state.
+
+If the author of the `Derived` class wants to allow construction via the
+base class constructor then this support must be added explicitly by
+writing forwarding constructors:
+
+```D
+class Base
+{
+    int sum;
+
+    this ( int x, int y )
+    {
+        this.sum = x + y;
+    }
+}
+
+class Derived : Base
+{
+    this ( )
+    {
+        super(5, 5);
+    }
+
+    // forwarding constructor
+    this ( int x, int y )
+    {
+        super(x, y);
+    }
+}
+
+void main ( )
+{
+    auto d1 = new Derived();  // ok
+    assert(d1.sum == 10);
+
+    auto d2 = new Derived(10, 10);  // ok
+    assert(d2.sum == 20);
+}
+```
+
+## Proposed features
+
+### 1. Add implicit constructor inheritance for derived classes with no constructors
+
+If a derived class does not define any constructors of its own then that class
+will automatically inherit all base class constructors.
+
+Example:
+
+```D
+class ParseException : Exception
+{
+    // inherits all of Exception's constructors
+}
+
+void parseArgs ( string[] args )
+{
+    if (args < 2)
+    {
+        throw new ParseException("Expected at least one argument");
+    }
+}
+
+int main ( string[] args )
+{
+    try
+    {
+        parseArgs(args);
+
+        return 0;
+    }
+    catch (ParseException ex)
+    {
+        stderr.writefln("Parsing error: %s", ex.msg);
+    }
+    catch (Exception ex)
+    {
+        stderr.writefln("Unexpected error: %s", ex.msg);
+    }
+
+    return 1;
+}
+```
+
+**Rationale**: If a derived class does not define any of its own constructors then
+there is no danger of leaving the object in an unexpected state.
+
+Side-note: Forwarding constructors are most prominent in exception class hierarchies.
+
+### 2. Add language support for inheriting base class constructors for derived classes with constructors
+
+If the author of the derived class defines its own set of constructors but also
+wants to allow construction via the base class'es constructors then the
+new constructor inheritance syntax can be used for this purpose:
+
+```D
+enum ErrorCode
+{
+    FileNotFound
+}
+
+class FileException : Exception
+{
+    /// introduce overload which takes an error code
+    this ( ErrorCode error_code, string file = __FILE__, string line = __LINE__ )
+    {
+        super(error_code.to!string, file, line);
+    }
+
+    /// inherit base class constructors
+    alias super.this this;
```

Comment:
It's sort of odd to use `alias super this;` because it looks like subtyping, but on the other hand the derived object **is** a base class object already (implicitly converts to base).

I think I'll simply list a couple of proposed syntaxes and then let the community decide on the best choice.